discojs/src/dataset/dataset.ts

```diff
@@ -174,4 +185,62 @@ export class Dataset<T> implements AsyncIterable<T> {
     for await (const _ of this) ret++;
     return ret;
   }
+
+  /** Try to keep generated elements to avoid recomputing
+   *
+   * Drops everything when memory pressure is applied.
+   */
+  cached(): Dataset<T> {
+    return new CachingDataset(this.#content);
+  }
+}
+
+/**
+ * Avoid recomputing the parent dataset, without hogging memory
+ *
+ * As dataset operations can be time-consuming, this keeps a weak reference to
+ * the generated elements so that a second iteration might yield theses directly.
+ **/
+class CachingDataset<T> extends Dataset<T> {
+  // potential reference to all elements
+  // tristate: undefined == empty, [false, _] == filling, [true, _] == filled
+  #cache = new WeakRef<[filled: boolean, List<T>]>([false, List()]);
+
+  override [Symbol.asyncIterator](): AsyncIterator<T> {
+    const cached = this.#cache.deref();
+
+    if (cached !== undefined && cached[0]) {
+      debug("valid cache, reading from it");
+
+      // eslint-disable-next-line @typescript-eslint/require-await
+      return (async function* () {
+        yield* cached[1];
+      })();
+    }
+
+    debug("cache invalid, recomputing");
```

Comment:
```suggestion
    debug("cache invalid, reading from dataset");
```
Nitpick: can we tweak this debug statement a bit? I found it a bit misleading when it's the first time calling the dataset to see "recomputing" as if it had already been computed before but failed.